core/drivers/i2c/atmel_i2c.c

```diff
@@ -0,0 +1,337 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2021 Microchip
+ */
+
+#include <assert.h>
+#include <drivers/clk.h>
+#include <drivers/clk_dt.h>
+#include <drivers/i2c.h>
+#include <drivers/pinctrl.h>
+#include <io.h>
+#include <kernel/dt_driver.h>
+#include <libfdt.h>
+#include <matrix.h>
+#include <platform_config.h>
+#include <string.h>
+
+#define TWIHS_CR		0x0
+#define TWIHS_CR_SWRST		BIT(7)
+#define TWIHS_CR_SVDIS		BIT(5)
+#define TWIHS_CR_MSEN		BIT(2)
+#define TWIHS_CR_STOP		BIT(1)
+#define TWIHS_CR_START		BIT(0)
+
+#define TWIHS_MMR		0x4
+#define TWIHS_MMR_DADR_SHIFT	16
+#define TWIHS_MMR_DADR_MASK	0x7F
+#define TWIHS_MMR_MREAD		BIT32(12)
+
+#define TWIHS_CWGR		0x10
+#define TWIHS_CWGR_HOLD_SHIFT	24
+#define TWIHS_CWGR_HOLD_MAX	0x1F
+#define TWIHS_CWGR_CKDIV_SHIFT	16
+#define TWIHS_CWGR_CKDIV_MAX	0x7
+#define TWIHS_CWGR_CHDIV_SHIFT	8
+
+#define TWIHS_CKSRC	BIT32(20)
+
+#define TWIHS_SR	0x20
+#define TWIHS_SR_NACK	BIT32(8)
+#define TWIHS_SR_TXRDY	BIT32(2)
+#define TWIHS_SR_RXRDY	BIT32(1)
+#define TWIHS_SR_TXCOMP	BIT32(0)
+
+#define TWIHS_RHR	0x30
+#define TWIHS_THR	0x34
+
+#define TWIHS_WPMR		0xE4
+#define TWIHS_WPMR_WPKEY	SHIFT_U32(0x545749, 8)
+
+#define I2C_BUS_FREQ	400000
+
+struct atmel_i2c {
+	uint32_t sda_hold_time;
+	vaddr_t base;
+	struct clk *clk;
+};
+
+static TEE_Result atmel_i2c_send_one_byte(struct atmel_i2c *i2c, uint8_t byte)
+{
+	uint32_t sr = 0;
+
+	io_write32(i2c->base + TWIHS_THR, byte);
+
+	while (true) {
+		sr = io_read32(i2c->base + TWIHS_SR);
+		if (sr & TWIHS_SR_NACK) {
+			EMSG("I2C received NACK while writing !");
+			return TEE_ERROR_GENERIC;
+		}
+		if (sr & TWIHS_SR_TXRDY)
+			break;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static void atmel_i2c_wait_txcomp(struct atmel_i2c *i2c)
+{
+	uint32_t sr = 0;
+
+	while (true) {
+		sr = io_read32(i2c->base + TWIHS_SR);
+		if (sr & TWIHS_SR_TXCOMP)
+			return;
```

Comment:
timeout?

Reply:
Indeed, a timeout would be nice, but for the moment, this platform does not have the ARM architected timers so timeout_elapsed can not be used. I'll need to come up with another way to retrieve time (probably adding a new way to retrieve that from the TCB peripheral on the sama5d2).